pretyping/coercion.ml

```diff
@@ -448,29 +458,41 @@ let mu env sigma t =
       | None -> sigma, (None, v, IdCoe)
   in aux t
 
+let unify_product env sigma j =
+  let t = whd_all env sigma j.uj_type in
+  match EConstr.kind sigma t with
+  | Prod _ -> Inl sigma
+  | Evar ev ->
+    let sigma, _ = Evardefine.define_evar_as_product env sigma ev in
+    Inl sigma
+  | _ ->
+    let sigma, (domain_hole, _) = Evarutil.new_type_evar env sigma Evd.univ_flexible in
+    let env' = push_rel Context.(Rel.Declaration.LocalAssum (anonR, domain_hole)) env in
+    let sigma, (codomain_hole, _) = Evarutil.new_type_evar env' sigma Evd.univ_flexible in
```

Comment:
The bug at https://github.com/coq/coq/pull/12383#issuecomment-889017482 was the lack of this push_rel, so the codomain indeed had an ill-typed evar instance.

Reply:
Ah, that makes sense! Thanks!